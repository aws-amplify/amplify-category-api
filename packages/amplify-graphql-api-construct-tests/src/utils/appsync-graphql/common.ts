import { graphqlRequest } from '../../graphql-request';

export interface AppSyncGraphqlResponse<T> {
  statusCode: number;
  body: {
    data?: T;
    errors?: [any];
  };
}

export interface OperationAuthInputAccessToken {
  accessToken: string;
}

export interface OperationAuthInputApiKey {
  apiKey: string;
}

export const isOperationAuthInputAccessToken = (obj: any): obj is OperationAuthInputAccessToken => {
  return typeof (obj as OperationAuthInputAccessToken).accessToken === 'string';
};

export const isOperationAuthInputApiKey = (obj: any): obj is OperationAuthInputApiKey => {
  return typeof (obj as OperationAuthInputApiKey).apiKey === 'string';
};

export const doAppSyncGraphqlOperation = async (input: any): Promise<any> => {
  const { apiEndpoint, auth, query, variables } = input;
  const headers = {
    'Content-Type': 'application/json',
  };
  if (isOperationAuthInputAccessToken(auth)) {
    headers['Authorization'] = auth.accessToken;
  } else if (isOperationAuthInputApiKey(auth)) {
    headers['x-api-key'] = auth.apiKey;
  } else {
    throw new Error(`Unknown auth type: ${auth}`);
  }

  const payload: any = {
    query: query,
  };

  if (variables) {
    payload.variables = variables;
  }

  const result = await graphqlRequest(apiEndpoint, {
    method: 'POST',
    headers,
    body: JSON.stringify(payload),
  });

  return result;
};
