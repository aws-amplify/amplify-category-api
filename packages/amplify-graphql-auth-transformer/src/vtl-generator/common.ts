import { and, bool, equals, Expression, methodCall, qref, ref, str } from 'graphql-mapping-template';
import { IAM_AUTH_TYPE } from '../utils';

// note in the resolver that operation is protected by auth
export const setHasAuthExpression: Expression = qref(methodCall(ref('ctx.stash.put'), str('hasAuth'), bool(true)));

// a logical condition which returns true if request is made with non-Cognito IAM principal
export const isNonCognitoIAMPrincipal = and([
  equals(ref('util.authType()'), str(IAM_AUTH_TYPE)),
  methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityPoolId')),
  methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityId')),
]);
