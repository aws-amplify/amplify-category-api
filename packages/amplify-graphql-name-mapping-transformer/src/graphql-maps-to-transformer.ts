import { TransformerPluginBase, isSqlModel, isDynamoDbModel } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerPreProcessContextProvider,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { MapsToDirective } from '@aws-amplify/graphql-directives';
import { ObjectTypeDefinitionNode, DirectiveNode, ObjectTypeExtensionNode } from 'graphql';
import { createMappingLambda } from './field-mapping-lambda';
import { attachFilterAndConditionInputMappingSlot, attachInputMappingSlot, attachResponseMappingSlot } from './field-mapping-resolvers';
import { shouldBeAppliedToModel, getMappedName, updateTypeMapping, setTypeMappingInSchema } from './graphql-name-mapping';

export class MapsToTransformer extends TransformerPluginBase {
  constructor() {
    super('amplify-maps-to-transformer', MapsToDirective.definition, TransformerPluginType.GENERIC);
  }

  /**
   * During the AST tree walking, the mapsTo transformer registers any renamed models with the ctx.resourceHelper.
   */
  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider): void => {
    shouldBeAppliedToModel(definition, MapsToDirective.name);
    shouldBeAppliedToDDBModels(definition, ctx as TransformerContextProvider);
    const modelName = definition.name.value;
    const mappedName = getMappedName(definition, directive, MapsToDirective.name, ctx.inputDocument);
    updateTypeMapping(modelName, mappedName, ctx.resourceHelper.setModelNameMapping);
  };

  /**
   * Run pre-mutation steps on the schema to support mapsTo
   * @param context The pre-processing context for the transformer, used to store type mappings
   */
  preMutateSchema = (context: TransformerPreProcessContextProvider): void => {
    setTypeMappingInSchema(context, MapsToDirective.name);
  };

  /**
   * During the generateResolvers step, the mapsTo transformer reads all of the model field mappings from the resourceHelper and generates
   * VTL to map the current field names to the original field names
   */
  after = (context: TransformerContextProvider): void => {
    context.resourceHelper.getModelFieldMapKeys().forEach((modelName) => {
      if (isSqlModel(context, modelName)) {
        return;
      }
      const modelFieldMap = context.resourceHelper.getModelFieldMap(modelName);
      if (!modelFieldMap.getMappedFields().length) {
        return;
      }
      const lambdaDataSource = createMappingLambda(context.api.host, context.stackManager);
      modelFieldMap.getResolverReferences().forEach(({ typeName, fieldName, isList }) => {
        const resolver = context.resolvers.getResolver(typeName, fieldName);
        if (!resolver) {
          return;
        }
        if (typeName === 'Mutation') {
          attachInputMappingSlot({
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
          });
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList: false,
          });
        } else {
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList,
          });
        }
      });
    });
  };
}

export const shouldBeAppliedToDDBModels = (
  definition: ObjectTypeDefinitionNode | ObjectTypeExtensionNode,
  ctx: TransformerContextProvider,
): void => {
  const modelName = definition.name.value;
  if (!isDynamoDbModel(ctx, modelName)) {
    throw new Error(`${MapsToDirective.name} is only supported on DynamoDB models. ${modelName} is not a DDB model.`);
  }
};
