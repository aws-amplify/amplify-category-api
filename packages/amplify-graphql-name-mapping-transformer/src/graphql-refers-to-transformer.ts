import { TransformerPluginBase, isSqlModel, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerPreProcessContextProvider,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { RefersToDirective } from '@aws-amplify/graphql-directives';
import {
  ObjectTypeDefinitionNode,
  DirectiveNode,
  FieldDefinitionNode,
  ObjectTypeExtensionNode,
  InterfaceTypeDefinitionNode,
  Kind,
} from 'graphql';
import {
  shouldBeAppliedToModel,
  getMappedName,
  updateTypeMapping,
  setTypeMappingInSchema,
  getMappedFieldName,
  updateFieldMapping,
} from './graphql-name-mapping';
import { attachFieldMappingSlot } from './field-mapping-resolvers';

export class RefersToTransformer extends TransformerPluginBase {
  constructor() {
    super('amplify-refers-to-transformer', RefersToDirective.definition, TransformerPluginType.GENERIC);
  }

  /**
   * Register any renamed models with the ctx.resourceHelper.
   */
  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;
    shouldBeAppliedToModel(definition, RefersToDirective.name);
    shouldBeAppliedToRDSModels(definition, context);
    const modelName = definition.name.value;
    const mappedName = getMappedName(definition, directive, RefersToDirective.name, ctx.inputDocument);
    updateTypeMapping(modelName, mappedName, ctx.resourceHelper.setModelNameMapping);
  };

  /**
   * Register any renamed model fields with the ctx.resourceHelper.
   */
  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    ctx: TransformerSchemaVisitStepContextProvider,
  ): void => {
    if (parent.kind === Kind.INTERFACE_TYPE_DEFINITION) {
      throw new InvalidDirectiveError(
        `@refersTo directive cannot be placed on "${parent?.name?.value}" interface's ${definition?.name?.value} field.`,
      );
    }
    const context = ctx as TransformerContextProvider;
    const modelName = parent?.name?.value;
    shouldBeAppliedToModel(parent, RefersToDirective.name);
    shouldBeAppliedToRDSModels(parent, context);
    shouldNotBeOnRelationalField(definition, modelName);
    const mappedName = getMappedFieldName(parent, definition, directive, RefersToDirective.name);
    updateFieldMapping(modelName, definition?.name?.value, mappedName, ctx);
  };

  /**
   * During the generateResolvers step, the refersTo transformer reads all of the model field mappings from the resourceHelper and generates
   * VTL to store the field mappings in the resolver context stash
   */
  after = (context: TransformerContextProvider): void => {
    context.resourceHelper.getModelFieldMapKeys().forEach((modelName) => {
      if (!isSqlModel(context, modelName)) {
        return;
      }
      const modelFieldMap = context.resourceHelper.getModelFieldMap(modelName);
      if (!modelFieldMap.getMappedFields().length) {
        return;
      }

      modelFieldMap.getResolverReferences().forEach(({ typeName, fieldName, isList }) => {
        const resolver = context.resolvers.getResolver(typeName, fieldName);
        if (!resolver) {
          return;
        }
        attachFieldMappingSlot({
          resolver,
          resolverTypeName: typeName,
          resolverFieldName: fieldName,
          fieldMap: modelFieldMap.getMappedFields(),
        });
      });
    });
  };

  /**
   * Run pre-mutation steps on the schema to support refersTo
   * @param context The pre-processing context for the transformer, used to store type mappings
   */
  preMutateSchema = (context: TransformerPreProcessContextProvider): void => {
    setTypeMappingInSchema(context, RefersToDirective.name);
  };
}

export const shouldBeAppliedToRDSModels = (
  definition: ObjectTypeDefinitionNode | ObjectTypeExtensionNode,
  ctx: TransformerContextProvider,
): void => {
  const modelName = definition.name.value;
  if (!isSqlModel(ctx, modelName)) {
    throw new Error(`@${RefersToDirective.name} is only supported on SQL models. ${modelName} is not a SQL model.`);
  }
};

export const shouldNotBeOnRelationalField = (definition: FieldDefinitionNode, modelName: string): void => {
  const relationalDirectives = ['hasOne', 'hasMany', 'belongsTo', 'manyToMany'];
  if (definition?.directives?.some((directive) => relationalDirectives.includes(directive?.name?.value))) {
    throw new Error(
      `@${RefersToDirective.name} is not supported on "${definition?.name?.value}" relational field in "${modelName}" model.`,
    );
  }
};
