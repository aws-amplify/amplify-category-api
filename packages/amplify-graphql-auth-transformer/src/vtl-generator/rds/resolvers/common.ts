import {
  Expression,
  and,
  bool,
  compoundExpression,
  equals,
  forEach,
  ifElse,
  iff,
  list,
  methodCall,
  not,
  nul,
  obj,
  or,
  parens,
  printBlock,
  qref,
  raw,
  ref,
  set,
  str,
  toJson,
  ret,
} from 'graphql-mapping-template';
import { FieldDefinitionNode } from 'graphql';
import { OPERATION_KEY } from '@aws-amplify/graphql-model-transformer';
import { API_KEY_AUTH_TYPE, DEFAULT_UNIQUE_IDENTITY_CLAIM, IDENTITY_CLAIM_DELIMITER, RoleDefinition } from '../../../utils';
import { isNonCognitoIAMPrincipal, setHasAuthExpression } from '../../common';

/**
 * Generates default RDS expression
 */
export const generateDefaultRDSExpression = (iamAccessEnabled: boolean): string => {
  const exp = ref('util.unauthorized()');
  return printBlock('Default RDS Auth Resolver')(generateIAMAccessCheck(iamAccessEnabled, compoundExpression([exp, toJson(obj({}))])));
};

export const generateAuthRulesFromRoles = (
  roles: Array<RoleDefinition>,
  fields: Readonly<FieldDefinitionNode[]>,
  hasIdentityPoolId: boolean,
  hideAllowedFields = false,
): Expression[] => {
  const expressions = [];
  expressions.push(qref(methodCall(ref('ctx.stash.put'), str('hasAuth'), bool(true))), set(ref('authRules'), list([])));
  const fieldNames = fields.map((field) => field.name.value);
  expressions.push(getIamAdminRoleExpression());
  roles.forEach((role) => {
    expressions.push(convertAuthRoleToVtl(role, fieldNames, hasIdentityPoolId, hideAllowedFields));
  });
  return expressions;
};

const getIamAdminRoleExpression = (): Expression =>
  iff(
    and([ref('ctx.stash.adminRoles'), ref('ctx.stash.adminRoles.size() > 0')]),
    qref(
      methodCall(
        ref('authRules.add'),
        obj({
          provider: str('iam'),
          type: str('admin'),
          strict: bool(false),
          roles: ref('ctx.stash.adminRoles'),
        }),
      ),
    ),
  );

const convertAuthRoleToVtl = (
  role: RoleDefinition,
  fields: string[],
  hasIdentityPoolId: boolean,
  hideAllowedFields = false,
): Expression => {
  const allowedFields = getAllowedFields(role, fields).map((field) => str(field));
  const showAllowedFields = allowedFields && !hideAllowedFields && allowedFields.length > 0;
  // Api Key
  if (role.provider === 'apiKey') {
    return qref(
      methodCall(
        ref('authRules.add'),
        obj({
          type: str('public'),
          provider: str('apiKey'),
          ...(showAllowedFields && { allowedFields: list(allowedFields) }),
        }),
      ),
    );
  }

  // Lambda Authorizer
  else if (role.provider === 'function') {
    return qref(
      methodCall(
        ref('authRules.add'),
        obj({
          type: str('custom'),
          provider: str('function'),
          ...(showAllowedFields && { allowedFields: list(allowedFields) }),
        }),
      ),
    );
  }

  // IAM
  else if (role.provider === 'iam') {
    return qref(
      methodCall(
        ref('authRules.add'),
        obj({
          type: str(role.strategy),
          provider: str('iam'),
          roleArn: role.strategy === 'public' ? ref('ctx.stash.unauthRole') : ref('ctx.stash.authRole'),
          ...(role.strategy === 'private' && { cognitoIdentityPoolId: hasIdentityPoolId ? ref('ctx.stash.identityPoolId') : nul() }),
          ...(showAllowedFields && { allowedFields: list(allowedFields) }),
        }),
      ),
    );
  }

  // User Pools or OIDC
  else if (role.provider === 'userPools' || role.provider === 'oidc') {
    if (role.strategy === 'private') {
      return qref(
        methodCall(
          ref('authRules.add'),
          obj({
            type: str(role.strategy),
            provider: str(role.provider),
            ...(showAllowedFields && { allowedFields: list(allowedFields) }),
          }),
        ),
      );
    } else if (role.strategy === 'groups' && role.static) {
      return qref(
        methodCall(
          ref('authRules.add'),
          obj({
            type: str(role.strategy),
            provider: str(role.provider),
            allowedGroups: list([str(role.entity)]),
            groupClaim: str(role.claim),
            ...(showAllowedFields && { allowedFields: list(allowedFields) }),
          }),
        ),
      );
    } else if (role.strategy === 'owner') {
      const usingCognitoDefaultClaim = role.claim === DEFAULT_UNIQUE_IDENTITY_CLAIM && role.provider === 'userPools';
      return qref(
        methodCall(
          ref('authRules.add'),
          obj({
            type: str(role.strategy),
            provider: str(role.provider),
            ownerFieldName: str(role.entity),
            ownerFieldType: str(role.isEntityList ? 'string[]' : 'string'),
            ...(!usingCognitoDefaultClaim && { identityClaim: str(role.claim) }),
            ...(showAllowedFields && { allowedFields: list(allowedFields) }),
          }),
        ),
      );
    } else if (role.strategy === 'groups' && !role.static) {
      return qref(
        methodCall(
          ref('authRules.add'),
          obj({
            type: str(role.strategy),
            provider: str(role.provider),
            groupsFieldName: str(role.entity),
            groupsFieldType: str(role.isEntityList ? 'string[]' : 'string'),
            groupClaim: str(role.claim),
            ...(showAllowedFields && { allowedFields: list(allowedFields) }),
          }),
        ),
      );
    }
  }
  throw new Error(`Invalid Auth Rule: Unable to process ${JSON.stringify(role)}`);
};

const getAllowedFields = (role: RoleDefinition, fields: string[]): string[] => {
  if (role.allowedFields && role.allowedFields.length > 0) {
    return role.allowedFields;
  }
  return fields;
};

export const validateAuthResult = (): Expression => {
  return compoundExpression([
    iff(
      or([not(ref('authResult')), parens(and([ref('authResult'), not(ref('authResult.authorized'))]))]),
      methodCall(ref('util.unauthorized')),
    ),
  ]);
};

export const constructAuthFilter = (): Expression => {
  return iff(
    and([ref('authResult'), not(methodCall(ref('util.isNullOrEmpty'), ref('authResult.authFilter')))]),
    set(ref('ctx.stash.authFilter'), ref('authResult.authFilter')),
  );
};

export const constructAuthorizedInputStatement = (keyName: string): Expression =>
  iff(
    and([ref('authResult'), not(methodCall(ref('util.isNullOrEmpty'), ref('authResult.authorizedInput')))]),
    set(ref(keyName), ref('authResult.authorizedInput')),
  );

/**
 * Generates sandbox expression for field
 */
export const generateSandboxExpressionForField = (sandboxEnabled: boolean, genericIamAccessEnabled: boolean): string => {
  const expressions: Array<Expression> = [];
  if (sandboxEnabled) {
    expressions.push(iff(equals(methodCall(ref('util.authType')), str(API_KEY_AUTH_TYPE)), ret(toJson(obj({})))));
  }
  if (genericIamAccessEnabled) {
    expressions.push(iff(isNonCognitoIAMPrincipal, ret(toJson(obj({})))));
  }
  expressions.push(methodCall(ref('util.unauthorized')));
  return printBlock(`Sandbox Mode ${sandboxEnabled ? 'Enabled' : 'Disabled'}`)(compoundExpression(expressions));
};

export const emptyPayload = toJson(raw(JSON.stringify({ version: '2018-05-29', payload: {} })));

/**
 * Creates expression to deny field flag
 */
export const setDeniedFieldFlag = (operation: string, subscriptionsEnabled: boolean): string => {
  if (subscriptionsEnabled) {
    return printBlock('Check if subscriptions is protected')(
      compoundExpression([
        iff(
          equals(methodCall(ref('util.defaultIfNull'), methodCall(ref('ctx.source.get'), str(OPERATION_KEY)), nul()), str(operation)),
          qref(methodCall(ref('ctx.stash.put'), str('deniedField'), bool(true))),
        ),
      ]),
    );
  }
  return '';
};

/**
 * Creates field resolver for owner
 */
export const generateFieldResolverForOwner = (entity: string): string => {
  const expressions: Expression[] = [
    ifElse(
      methodCall(ref('util.isList'), ref(`ctx.source.${entity}`)),
      compoundExpression([
        set(ref('ownerEntitiesList'), list([])),
        set(ref(entity), ref(`ctx.source.${entity}`)),
        forEach(ref('entities'), ref(entity), [
          set(ref('ownerEntities'), ref(`entities.split("${IDENTITY_CLAIM_DELIMITER}")`)),
          set(ref('ownerEntitiesLastIdx'), raw('$ownerEntities.size() - 1')),
          set(ref('ownerEntitiesLast'), ref('ownerEntities[$ownerEntitiesLastIdx]')),
          qref(methodCall(ref('ownerEntitiesList.add'), ref('ownerEntitiesLast'))),
        ]),
        qref(methodCall(ref(`ctx.source.${entity}.put`), ref('ownerEntitiesList'))),
        toJson(ref('ownerEntitiesList')),
      ]),
      compoundExpression([
        set(ref('ownerEntities'), ref(`ctx.source.${entity}.split("${IDENTITY_CLAIM_DELIMITER}")`)),
        set(ref('ownerEntitiesLastIdx'), raw('$ownerEntities.size() - 1')),
        set(ref('ownerEntitiesLast'), ref('ownerEntities[$ownerEntitiesLastIdx]')),
        qref(methodCall(ref('ctx.source.put'), str(entity), ref('ownerEntitiesLast'))),
        toJson(ref(`ctx.source.${entity}`)),
      ]),
    ),
  ];

  return printBlock('Parse owner field auth for Get')(compoundExpression(expressions));
};

/**
 * If generic IAM access is enabled add check that bypasses rules evaluation for any IAM principal
 * that is not related to Cognito Identity Pool.
 */
export const generateIAMAccessCheck = (enableIamAccess: boolean, expression: Expression): Expression => {
  if (!enableIamAccess) {
    // No-op if generic IAM access is not enabled.
    return expression;
  }
  return ifElse(isNonCognitoIAMPrincipal, compoundExpression([setHasAuthExpression, emptyPayload]), expression);
};
