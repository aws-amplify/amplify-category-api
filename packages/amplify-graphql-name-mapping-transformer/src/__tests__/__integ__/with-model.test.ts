import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { MapsToTransformer } from '@aws-amplify/graphql-maps-to-transformer';
import { DDB_DB_TYPE, MYSQL_DB_TYPE, DBType } from '@aws-amplify/graphql-transformer-core';
import { constructModelToDataSourceMap, testTableNameMapping, testColumnNameMapping } from './common';
import { RefersToTransformer } from '../../graphql-refers-to-transformer';

const transformSchema = (schema: string, dbType: DBType) => {
  return testTransform({
    schema,
    transformers: [new ModelTransformer(), new MapsToTransformer(), new RefersToTransformer()],
    modelToDatasourceMap: constructModelToDataSourceMap(['Todo'], dbType),
    transformParameters: {
      sandboxModeEnabled: true,
    },
  });
};

describe('@mapsTo directive on model type', () => {
  it('generates table name with mapped name', () => {
    const basicSchema = /* GraphQL */ `
      type Todo @model @mapsTo(name: "Task") {
        id: ID!
        title: String!
      }
    `;
    const out = transformSchema(basicSchema, DDB_DB_TYPE);
    expect(out.stacks.Task.Resources!.TaskTable!.Properties.TableName).toMatchInlineSnapshot(`
      Object {
        "Fn::Join": Array [
          "",
          Array [
            "Task-",
            Object {
              "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
            },
            "-",
            Object {
              "Ref": "referencetotransformerrootstackenv10C5A902Ref",
            },
          ],
        ],
      }
    `);
    expect(out.stacks.Task.Outputs!.GetAttTaskTableName).toMatchInlineSnapshot(`
      Object {
        "Description": "Your DynamoDB table name.",
        "Export": Object {
          "Name": Object {
            "Fn::Join": Array [
              ":",
              Array [
                Object {
                  "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
                },
                "GetAtt:TaskTable:Name",
              ],
            ],
          },
        },
        "Value": Object {
          "Ref": "TaskTable",
        },
      }
    `);
  });
});

describe('@refersTo with RDS Models', () => {
  it('model table names are mapped', () => {
    const basicSchema = /* GraphQL */ `
      type Todo @model @refersTo(name: "Task") {
        id: ID!
        title: String!
      }
    `;
    const out = transformSchema(basicSchema, MYSQL_DB_TYPE);
    testTableNameMapping('Todo', 'Task', out);
  });

  it('model field names are mapped', () => {
    const basicSchema = /* GraphQL */ `
      type Todo @model {
        id: ID!
        title: String! @refersTo(name: "description")
      }
    `;
    const out = transformSchema(basicSchema, MYSQL_DB_TYPE);
    testColumnNameMapping('Todo', out);
  });
});
