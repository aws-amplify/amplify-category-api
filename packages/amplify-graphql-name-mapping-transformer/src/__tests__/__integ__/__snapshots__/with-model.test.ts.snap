// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@refersTo with SQL Models model field names are mapped 1`] = `
"#set( $fieldMap = {\\"title\\":\\"description\\"} )
#set( $ctx.stash.fieldMap = $fieldMap )
$util.toJson({})"
`;

exports[`@refersTo with SQL Models model field names are mapped 2`] = `
"#set( $fieldMap = {\\"title\\":\\"description\\"} )
#set( $ctx.stash.fieldMap = $fieldMap )
$util.toJson({})"
`;

exports[`@refersTo with SQL Models model field names are mapped 3`] = `
"#set( $fieldMap = {\\"title\\":\\"description\\"} )
#set( $ctx.stash.fieldMap = $fieldMap )
$util.toJson({})"
`;

exports[`@refersTo with SQL Models model field names are mapped 4`] = `
"#set( $fieldMap = {\\"title\\":\\"description\\"} )
#set( $ctx.stash.fieldMap = $fieldMap )
$util.toJson({})"
`;

exports[`@refersTo with SQL Models model field names are mapped 5`] = `
"#set( $fieldMap = {\\"title\\":\\"description\\"} )
#set( $ctx.stash.fieldMap = $fieldMap )
$util.toJson({})"
`;

exports[`@refersTo with SQL Models model table names are mapped 1`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.table = \\"Task\\" )
#set( $lambdaInput.args = {} )
#set( $lambdaInput.operation = \\"CREATE\\" )
#set( $lambdaInput.operationName = \\"createTodo\\" )
#set( $lambdaInput.args.metadata = {} )
#set( $lambdaInput.args.metadata.keys = [] )
#set( $lambdaInput.args.metadata.nonScalarFields = [] )
#set( $lambdaInput.args.metadata.arrayFields = [] )
#set( $lambdaInput.args.metadata.fieldMap = {} )
$util.qr($lambdaInput.args.metadata.fieldMap.putAll($util.defaultIfNull($context.stash.fieldMap, {})))
$util.qr($lambdaInput.args.metadata.keys.addAll($util.defaultIfNull($ctx.stash.keys, [])))
## Set the default values to put request **
#set( $lambdaInput.args.input = $util.defaultIfNull($ctx.stash.defaultValues, {}) )
## copy the values from input **
$util.qr($lambdaInput.args.input.putAll($util.defaultIfNull($context.arguments.input, {})))
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`@refersTo with SQL Models model table names are mapped 2`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.table = \\"Task\\" )
#set( $lambdaInput.args = {} )
#set( $lambdaInput.operation = \\"UPDATE\\" )
#set( $lambdaInput.operationName = \\"updateTodo\\" )
#set( $lambdaInput.args.metadata = {} )
#set( $lambdaInput.args.metadata.keys = [] )
#set( $lambdaInput.args.metadata.nonScalarFields = [] )
#set( $lambdaInput.args.metadata.arrayFields = [] )
#set( $lambdaInput.args.metadata.fieldMap = {} )
$util.qr($lambdaInput.args.metadata.fieldMap.putAll($util.defaultIfNull($context.stash.fieldMap, {})))
$util.qr($lambdaInput.args.metadata.keys.addAll($util.defaultIfNull($ctx.stash.keys, [])))
## Set the default values to put request **
#set( $lambdaInput.args.input = $util.defaultIfNull($ctx.stash.defaultValues, {}) )
## copy the values from input **
$util.qr($lambdaInput.args.input.putAll($util.defaultIfNull($context.arguments.input, {})))
#set( $lambdaInput.args.condition = $util.defaultIfNull($context.arguments.condition, {}) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`@refersTo with SQL Models model table names are mapped 3`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.table = \\"Task\\" )
#set( $lambdaInput.args = $context.arguments )
#set( $lambdaInput.operation = \\"DELETE\\" )
#set( $lambdaInput.operationName = \\"deleteTodo\\" )
#set( $lambdaInput.args.metadata = {} )
#set( $lambdaInput.args.metadata.keys = [] )
#set( $lambdaInput.args.metadata.nonScalarFields = [] )
#set( $lambdaInput.args.metadata.arrayFields = [] )
#set( $lambdaInput.args.metadata.fieldMap = {} )
$util.qr($lambdaInput.args.metadata.fieldMap.putAll($util.defaultIfNull($context.stash.fieldMap, {})))
$util.qr($lambdaInput.args.metadata.keys.addAll($util.defaultIfNull($ctx.stash.keys, [])))
#set( $lambdaInput.args.condition = $util.defaultIfNull($context.arguments.condition, {}) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`@refersTo with SQL Models model table names are mapped 4`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.args = {} )
#set( $lambdaInput.table = \\"Task\\" )
#set( $lambdaInput.operation = \\"GET\\" )
#set( $lambdaInput.operationName = \\"getTodo\\" )
#set( $lambdaInput.args.metadata = {} )
#set( $lambdaInput.args.metadata.keys = [] )
#set( $lambdaInput.args.metadata.authFilter = {} )
#set( $lambdaInput.args.metadata.nonScalarFields = [] )
#set( $lambdaInput.args.metadata.arrayFields = [] )
#set( $lambdaInput.args.metadata.fieldMap = {} )
$util.qr($lambdaInput.args.metadata.fieldMap.putAll($util.defaultIfNull($context.stash.fieldMap, {})))
$util.qr($lambdaInput.args.metadata.keys.addAll($util.defaultIfNull($ctx.stash.keys, [])))
#set( $lambdaInput.args.input = $util.defaultIfNull($ctx.stash.defaultValues, {}) )
$util.qr($lambdaInput.args.input.putAll($util.defaultIfNull($context.arguments, {})))
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`@refersTo with SQL Models model table names are mapped 5`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.args = {} )
#set( $lambdaInput.table = \\"Task\\" )
#set( $lambdaInput.operation = \\"LIST\\" )
#set( $lambdaInput.operationName = \\"listTodos\\" )
#set( $lambdaInput.args.metadata = {} )
#set( $lambdaInput.args.metadata.keys = [] )
#if( !$util.isNullOrEmpty($ctx.stash.authFilter) )
  #set( $lambdaInput.args.metadata.authFilter = $ctx.stash.authFilter )
#end
#set( $lambdaInput.args.metadata.nonScalarFields = [] )
#set( $lambdaInput.args.metadata.arrayFields = [] )
#set( $lambdaInput.args.metadata.fieldMap = {} )
$util.qr($lambdaInput.args.metadata.fieldMap.putAll($util.defaultIfNull($context.stash.fieldMap, {})))
$util.qr($lambdaInput.args.metadata.keys.addAll($util.defaultIfNull($ctx.stash.keys, [])))
#set( $lambdaInput.args.input = $util.defaultIfNull($ctx.stash.defaultValues, {}) )
$util.qr($lambdaInput.args.putAll($util.defaultIfNull($context.arguments, {})))
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;
